### DISCUSSION TITLE
Mutations that also provide the new state

### DISCUSSION THREAD

**Discussion by bvdoord** [reduxjs/redux-toolkit#1169] [Q&A] [answered]
Hello everyone,

I'm working with an API that provides endpoints for mutations that also send back the new state of the mutated entity as the response. While the regular flow of invalidating the cache works just fine, I'm wondering it it's possible to combine a mutation with a query.

Let's say I have a query for a Product, with a tag (of type `"product"` with the product id.)
I also have a mutation to change the Product name. This endpoint returns the updated product entirely. I could invalidate the tag, and let the above query fire a second time (as normal.) But this feels redundant, because I already have the updated Product.

Currently mutations can't provide tags. Is this something that's planned, or should I look into something like the optimistic updates?

**Comment by phryneas** [comment, marked as answer]
> Currently mutations can't provide tags. Is this something that's planned, or should I look into something like the optimistic updates?

That wouldn't change anything with your problem I think.  
That something "provides a tag" only means that it will be re-fetched if that tag is invalidated - a tag is only an indicator to the system to trigger a re-fetch, not holding any data. Since we assume that sending a mutation actually initiates a *change in the database* on the server, we will never do that automatically - that always has to be implicit. (If your mutation only fetches data without triggering a change in the backend, it should just not be a mutation, but a query.)

If you want to use the return value from your mutation to update cache values for queries, you are correct that you can use the same mechanisms you'd use for optimistic updates for that. You'd just wait for the result and *then* use the result value to update the cache value of your query, instead of doing it at query initialization and rolling back on error.

**Reply by bdugg-91** [reply]
Sorry for the thread necromancy, but is there any place to do further reading for this sort of operation? 

I'm currently running into the same type of dilemma wherein I'd like to execute a mutation, and then use the response object to update the cached data for the appropriate tag.

**Reply by markerikson** [reply]
Yep, see https://redux-toolkit.js.org/rtk-query/usage/manual-cache-updates#pessimistic-updates
